### ISSUE TITLE
SRCNN degrades an extracted patch; the authors degrade the image and then extract

### ISSUE THREAD

**Issue by YeapJieShen** [YeapJieShen/single-image-super-resolution#245]
## The defect

SRCNN's training path degrades an **extracted HR patch**. The authors degrade the **whole
image** and extract patches from the result.

Ours, in `datasets/srcnn.py::TrainDataset.__getitem__`:

```python
hr_subimg  = <33x33 slice of the HR image>
lr_subimg  = _degrade(hr_subimg, self.scale)     # degrade the PATCH
```

Theirs, in the released `demo_SR.m` / `generate_train.m` order:

```matlab
im_gnd = modcrop(im, up_scale);
im_l   = imresize(im_gnd, 1/up_scale, 'bicubic');   % degrade the IMAGE
im_b   = imresize(im_l,   up_scale,   'bicubic');
% ... only then are 33x33 sub-images taken from im_gnd / im_b
```

These are not the same operation. Bicubic interpolation's kernel support extends past the
patch boundary, so degrading a 33x33 crop resolves those taps against the crop's edge
handling instead of against the real neighbouring pixels.

## Measured

Smooth, image-like content (noise exaggerates this heavily — on random input the same
comparison reads max 36–59 levels, which is not representative):

| scale | whole patch max\|diff\| | mean | 4 px border mean | interior mean |
|---:|---:|---:|---:|---:|
| 2 | 3 | 0.519 | 0.663 | 0.424 |
| **3** | **5** | 0.333 | **0.831** | **0.000** |
| 4 | 8 | 0.904 | 1.574 | 0.450 |

**At scale 3 the interior is bit-identical and the entire difference is the border.** That is
the signature of the kernel reaching outside the patch and nothing else. At scales 2 and 4 the
interior differs too, because `sub_img_size = 33` is not a multiple of those scales, so the LR
sampling grid also shifts.

A 4 px border is **43%** of a 33x33 patch's pixels.

## Why it is worth filing rather than shrugging at

The differences are small in absolute terms, and this is deliberately filed as a fidelity gap
rather than a correctness bug. But this is a paper-reproduction framework, and the standing
rule is that a figure is comparable only to one computed the same way. Here the *training
inputs* are not produced the way the authors produced theirs — so a reproduction gap has a
candidate explanation that is not the model.

It also means `sub_img_size` and `scale` are silently coupled: `33 % 3 == 0` makes scale 3
much better behaved than 2 or 4, and nothing anywhere says so.

## Relationship to the modcrop issue

A separate issue covers SRCNN *validation* skipping modcrop. That one is about **which pixels
exist**; this one is about **when the degradation is applied relative to patching**. Fixing
the first does not fix this — after it, train and eval agree on the modcrop convention and
still disagree on this axis.

## Scope note

Not free. Degrading whole images at load time, or caching an LR plane alongside the HR one,
changes the data path this project measured at 98.94% decode-dominated — so the fix needs a
throughput measurement, not just a correctness one.

## Acceptance

- [ ] SRCNN training patches are cut from a degraded whole image, matching the released order
- [ ] A test on a patch whose border demonstrably differs today, proven RED first
- [ ] `sub_img_size` / `scale` coupling is either removed or documented where it is chosen
- [ ] The data path's throughput is measured before and after on this machine

**Comment by YeapJieShen** [comment]
Reopened. This was **not** fixed, and nothing here has changed.

It was closed automatically when the modcrop PR merged. That PR only *referenced* this issue —
to say the divergence was measured and deliberately left — and carried no closing keyword for
it. The close came from project automation reacting to the cross-reference, not from any code
change.

The measured difference stands unchanged: at x3 the patch interior is bit-identical and the
entire difference is the 4 px border, which is 43% of a 33x33 patch's pixels.

Still needs what the acceptance list says, in particular the throughput measurement — degrading
whole images at load time, or caching an LR plane, changes a data path measured at 98.94%
decode-dominated.

**Comment by YeapJieShen** [comment]
## Throughput before/after — it is a speedup, not a cost

This issue's scope note assumed matching the reference order would cost
throughput. It does the opposite. The per-item bicubic resize was what remained
after the LMDB cache removed the decode, and slicing a pre-derived plane deletes
it.

Dataset-level `__getitem__` — no model, no GPU, no DataLoader workers, since a
pool would measure the pool. 50 warmup, 400 samples, medians. On mains, GPU
idle, nothing else running, real DIV2K training images.

| dataset | before | after | change |
|---|---:|---:|---|
| SRResNet x4, 96px crop | 10.1086 ms/item — 98.9 items/s | **0.8316 ms/item — 1,202.4 items/s** | **12.2x faster** |
| SRCNN x3, 33px patch | 0.3891 ms/item — 2,569.7 items/s | **0.0810 ms/item — 12,342.8 items/s** | **4.8x faster** |

SRResNet's 10.1 ms/item was a real bottleneck: at batch 16 that is ~162 ms of
data work per batch to hide behind the step.

**What it costs instead is disk, once per (directory, kind, scale).** The LR
plane at x4 is **401 MB**, built in **62.8 s**. SRCNN's pre-upsampled plane is
HR-sized by construction: **6.3 GB**, built in **101.9 s**. The raw-HR cache is
untouched and still shared across architectures.

**Comment by YeapJieShen** [comment]
## Correction — my earlier throughput comment on this issue was measured wrong

**Retracting the "12.2x / 4.8x speedup" I posted above.** It came from a
dataset-level microbenchmark that does not represent a training loader, and for
SRCNN it was wrong in **direction**, not merely magnitude.

### What the benchmark got wrong

- It walked **consecutive** indices. For SRCNN that reads neighbouring patches of
  the same image, which is far more page-cache-friendly than anything a real run
  does — a real loader shuffles across every patch.
- 2000 samples out of 11 million patches touch roughly 13 MB. After warmup it
  was measuring cache hits, not the steady-state miss rate.
- One process, where a real run has 16 workers competing for the same page cache.

### What a real training loop says

400 real training steps, no validation, on mains, GPU otherwise idle:

| | derived plane | working set | before | after |
|---|---:|---:|---:|---:|
| **SRResNet x4** | 401 MB | 6.7 GB | 22.25 it/s | **23.51 it/s** |
| **SRCNN x3** | 6.3 GB | **12.6 GB** | 21.60 it/s | **1.02 it/s** |

**A 21x regression for SRCNN.** The box has 12 GB of RAM. SRCNN is
pre-upsampled, so its degraded plane is HR-sized rather than 1/scale^2 of it;
caching it doubles the working set past RAM and a shuffled loader then misses on
nearly every item. Confirmed at the time by the symptom, not just the rate: GPU
at 13%, 36% iowait, 992 MB/s sustained disk read.

SRResNet is unaffected — its plane is 1/16 the pixels, the working set still
fits, and it gains slightly.

### What changed as a result

SRCNN **keeps degrading each extracted patch**, and this fidelity gap stays open.
SRResNet takes the whole-image order, verified byte-exact against DIV2K's own
distributed LR. `derived_kind=None` is now an explicit, documented opt-out on the
shared base rather than an omission.

After the revert SRCNN measures **34.23 it/s** — above the original 21.60,
because the 6.3 GB plane is no longer competing for page cache.

### What would close this properly

Caching the small LR plane (~0.7 GB at x3) and upsampling only the needed window
at load, with the output grid aligned to the whole-image resize. That keeps the
working set inside RAM and preserves byte-exactness, but it needs an `imresize`
variant that resizes a window at a global grid offset — real work, and it must
re-prove byte-identical output.

Alternatively this is simply a small-box problem: at 16 GB or more the cached
plane fits and the original approach is fine as written.
